ML-297 Initial SupportVectorMachines Bundle

timothyesler opened the pull request:
@johnholt I've separated this into two commits in the hope it makes the review easier. The first was just for the migration of existing code from the old ECL-ML repo, and the second is the changes I've made.

Commit:
ML-297 SVM Bundle using the LibSVM bundle. Myriad interface.

Initial migration from ecl-ml to dedicated bundle, as per ML library restructure.

Commit:
ML-297 SVM Bundle

Initial commit of SupportVectorMachines bundle repo, as per ML library
restructure.

The following changes are included:

- Remove redundant code files.
- Add defaults for key RECORD structures.
- Implement problem-focused node-use optimization for grid search.
- Add javadoc style documentation for all user-facing functionality.
- Add 'walkthrough' examples exemplifying the use of the bundle for
training, predicting, testing, and tuning.
- Add function for generating human-readable SVM model summary.
- Add standard IClassification and IRegression interfaces (SVC.ecl and
SVR.ecl).
- Implement optional input data standardization (centering and scaling).
- Implement myriad interface.
- Add example heart_scale dataset in CSV and ECL-module forms (for
examples and validation).
- Add validation scripts to compare HPCC implementation of libSVM to
that of the e1071 package in R.
- Various code optimizations to ensure the myriad interface runs efficiently (mostly changes to model conversion functions)

Signed-off-by: Timothy Esler <timothy.esler@lexisnexisrisk.com>

johnholt commented on Converted.ecl:
```diff
@@ -25,7 +40,7 @@ EXPORT Converted := MODULE
       SELF := i;
     END;
-    rslt := JOIN(indy, dep, LEFT.rid=RIGHT.id, getD(LEFT,RIGHT),
+    rslt := JOIN(indy, dep, LEFT.rid=RIGHT.id AND LEFT.wi=RIGHT.wi, getD(LEFT,RIGHT),
```
The compiler should be smart enough, but just in case, it would be better to have LEFT.wi=RIGHT.wi first so that the sort and distribution is certain to be based on work item as the high order partition value.

johnholt commented on README.md:
```diff
@@ -1,2 +1,4 @@
 # SupportVectorMachines
 Support Vector Machine implementations.
+
```
Add a note that some distributions may have a development version, and that the development version is required if there is more than one version.

johnholt commented on predict.ecl:
```diff
@@ -1,6 +1,8 @@
 ﻿// Produce a data set of predictions for each model
-IMPORT $ as SVM;
-IMPORT $.LibSVM;
+IMPORT SupportVectorMachines as SVM;
```
It is usually better to use $ AS SVM if you are using Eclipse or VS Code.  Does the ECL IDE give you a problem if you don't use the directory name?

Commit:
ML-297 SVM Bundle

Response to pull request review.

The following updates are included:

- Removed a few files that had been renamed/deleted but somehow made their way back (e.g. svm_train.ecl is not used in preference for SVMTrain.ecl)
- Reordered id and wi in JOIN in Converted.ecl to ensure correct compiler interpretation
- Added note about development version in README
- Used $ throughout, instead of directory name

Signed-off-by: Timothy Esler <timothy.esler@lexisnexisrisk.com>

ghost commented:
@johnholt I've pushed another commit to address your comments, and also to remove a few old files that shouldn't have been in there.

Another thing I noticed is that when using the ToField MACRO in ML_Core, it will throw an error if the calling script has aliased ML_Core to anything other than ML_Core. As a result I have avoided using aliases for ML_Core in this repo.
